### ISSUE TITLE
Support overriding CRD type validation for structs

### ISSUE THREAD

**Issue by joelanford** [kubernetes-sigs/controller-tools#155]
Currently, it seems the validation type for structs is always assumed to be `object`.

https://github.com/kubernetes-sigs/controller-tools/blob/5752ac22027d963d0d0c4ccd9c4a92620ee0f921/pkg/internal/codegen/parse/crd.go#L436-L441

However, structs can be serialized into types other than `object`. Is there a way to override the type detection for structs? 

I'm trying to do something like the following:
```go
package status

// Conditions manages status conditions
type Conditions struct {
    conditions map[ConditionType]Condition
}

// MarshalJSON marshals Conditions into a JSON array 
func (c Conditions) MarshalJSON() ([]byte, error) {
    conditionList := []Condition{}
    for k, v := range c.conditions {
        conditionList = append(conditionList, v)
    }
    return json.Marshal(conditionList)
}
```

```go
package v1alpha

// +k8s:openapi-gen=true
type MyAppStatus struct {
    // When the CRD generator creates the validations, it sees that Conditions
    // is a struct, but since it is marshalled as a JSON list, the default type detection
    // doesn't work.
    //
    // Perhaps support for a new Type validation tag could be added?
    // +kubebuilder:validation:Type=array
    Conditions status.Conditions `json:"conditions"`
}
```

**Comment by negz** [comment]
I have a similar use case for primitive types, in which I'd like to define an 'enum' as an integer in my Go code, but have it marshal to a JSON string. This is handy because it allows a sane default for the zero type of an enum. For example:

```golang
// WorkloadState represents the state of a workload.
type WorkloadState int

// MarshalJSON returns a JSON representation of a WorkloadState.
func (s WorkloadState) MarshalJSON() ([]byte, error) {
	return json.Marshal(s.String())
}

// UnmarshalJSON returns a WorkloadState from its JSON representation.
func (s *WorkloadState) UnmarshalJSON(b []byte) error {
	var ss string
	if err := json.Unmarshal(b, &ss); err != nil {
		return err
	}
	switch ss {
	case WorkloadStateSynced.String():
		*s = WorkloadStateSynced
	default:
		*s = WorkloadStateUnknown
	}
	return nil
}

// Workload states.
const (
	WorkloadStateUnknown WorkloadState = iota
	WorkloadStateSynced
)

// WorkloadStatus represents the status of a workload.
type WorkloadStatus struct {
	// ...
	State   WorkloadState           `json:"state"`
}
```

https://github.com/kubernetes-sigs/controller-tools/blob/45a8321/pkg/internal/codegen/parse/crd.go#L285

Currently the controller-tools CRD generation code does not appear to support overriding the type detected by reflect. For my use case we could perhaps use the same notation/pattern as openapi-gen's [custom type definitions](https://github.com/kubernetes/kube-openapi/tree/master/pkg/generators#custom-openapi-type-definitions).

**Comment by negz** [comment]
/remove-lifecycle statle

This is still something that would make life easier as a user of controller-tools. FWIW I've also wanted to use this proposed functionality for the original purpose @joelanford mentioned (i.e. storing conditions as a map, but marshalling them to an array) as well as the case I mentioned.

**Comment by joelanford** [comment]
This was resolved in #200.
